### ISSUE TITLE
Curry Exercise 2: filterQs does not, in fact, filter Q's

### ISSUE THREAD

**Issue by endrin** [MostlyAdequate/mostly-adequate-guide#84]
When the `filterQs` function from Curry Exercise 2 is applied to test array (`['quick', 'camels', 'quarry', 'over', 'quails']`) it returns the same array unchanged.

The cause of this is the `match` function which is used by `filterQs`. It returns empty array when the regexp and string to which it is being applied does not match each other. Empty array is considered truthy, so the Ramda's `filter` does not filter anything.

I made provided test pass by changing the function to this:
`var filterQs = _.reject(_.compose(_.isEmpty, _.match(/q/i)));`

But this is probably not the best way to solve the exercise, as composition is subject of the next chapter.

As possible solution, the Ramda's `match` can be replaced with the `match` from the `support` module, so the exercise will look like this:

```javascript
// Exercise 2
//==============
// Refactor to remove all arguments by partially applying the functions
// Use the provided _match function

// LEAVE BE:
var _match = _.curry(function(what, x) { return x.match(what); });

// REFACTOR THIS ONE:
var filterQs = function(xs) {
  return _.filter(function(x){ return _match(/q/i, x);  }, xs);
};
```

**Comment by endrin** [comment]
As possible solution, the Ramda's `match` can be replaced with the `match` from the `support` module, so the exercise will look like this:

```javascript
// Exercise 2
//==============
// Refactor to remove all arguments by partially applying the functions
// Use the provided _match function

// LEAVE BE:
var _match = _.curry(function(what, x) { return x.match(what); });

var filterQs = function(xs) {
  return _.filter(function(x){ return _match(/q/i, x);  }, xs);
};
```

**Comment by DrBoolean** [comment]
Ahhh, thank you. There was a previous pull request to use ramda's version as it's explicitly required. The tests still passed since we only updated the book text. I need to work out a way of pointing to the book's definitions of things like `match` so people don't get confused.

**Comment by legomushroom** [comment]
I was confused by this one too and was about to open an issue. Just saying :)

**Comment by legomushroom** [comment]
There is codepen with the issue, if it will help somehow http://codepen.io/sol0mka/pen/eNqPPx?editors=001

**Comment by CraigDMcKenna** [comment]
I am confused, are you saying that using `R.match` solves this?  I seems to return an empty array as well.

**How about this?**
```var test = function(what, x) {
  return _.test(what);
};

var filter = function(f, xs) {
  return _.filter(f);
};

var hasQs = test(/q/g);

var filterQs = filter(hasQs);
```

**Comment by CraigDMcKenna** [comment]
I am confused, are you saying that using R.match solves this? I seems to return an empty array as well.

How about this?
```
var test = function(what, x) {
  return _.test(what);
};

var filter = function(f, xs) {
  return _.filter(f);
};

var hasQs = test(/q/ig);

var filterQs = filter(hasQs);
```

**Comment by KtorZ** [comment]
Hey!
Not sure the issue is still relevant here. 

Using the match and filter functions provided in the `support` library, the solution indicated is okay. 

```js
// --- In support.js
match = curry(function(what, x) {
    return x.match(what);
});

filter = curry(function(f, xs) {
    return xs.filter(f);
});

// --- Exercice
var filterQs = filter(match(/q/i))
```

which is relatively equivalent to your proposal though it does not require anything else but the native JavaScript functions.

@CraigDMcKenna Do you face any issue with the current solution ?

**Comment by CraigDMcKenna** [comment]
I was using JS Bin to complete the exercises and thus had no reference to the helper functions on the repo.  The flow of the chapter led me to _assume_ that I should write my own helper functions as in the preceding examples.  I also _assumed_ that `match()` would use [String.prototype.match()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match).

I solved it using [RegExp.prototype.test()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test).

Then I decided clone the repo to "peek" at the answers but first I ran the tests which also confused me because all but one test passed and I hadn't written any code yet so I commented out `require('../../support');`

`.match()` just doesn't seem like the best solution in either of the two situations. Granted, I did make some assumptions regarding it's use. I'm just not sure that the book is quite clear.  I do feel, however that I have learned much through this process.

As a side note: an explanation of `reduce` would be helpful to me also.  @KtorZ  can you point me to something a little more descriptive than the Ramda Docs?

**Comment by CraigDMcKenna** [comment]
Also, I guess I'm still not sure how the tests pass when the helper function uses `.match()` which returns an array instead of a Boolean to `.filter()` .

**Comment by CraigDMcKenna** [comment]
Also, I guess I'm still not sure how the tests pass when the `match()` helper function uses `.match()` which returns an array not a Boolean.

**Comment by KtorZ** [comment]
Running the test runner on the answer folder works fine. 
However, you're completely right about the wrong use of the `match` function whereas we should have used `test`. `match` is more versatile and therefore, most of the time, one also uses it to simply check whether the string can be matched or not (in case of match, the function returns an array otherwise, null which is a falsy value).

I'll have a look and will likely do a little PR about this :)

**Comment by CraigDMcKenna** [comment]
Maybe Ramda should change ;-)

**Comment by CraigDMcKenna** [comment]
I know this is closed but I just wanted to clarify something. My confusion came form not understanding that __.SameNameAsJSMethod()_ is not the same as the JavaScript method of the same name.

The book explained that Ramda has currying built in which lead me to believe that calling _R.whatever_ or __.whatever_ (as it's used in the book) was somehow currying native JavaScript Methods because there was no explanation of the libraries 'custom' functions.

For someone who has used a library like underscore or lodash before this may not be an issue.

In other words.
```
.match() // JavaScript method

_.match() // or
R.match() 
// is a custom functional implementation of .match() with the library (underscore, lodash, ramda...)
```
